src/serde/decimal.rs

```diff
@@ -0,0 +1,79 @@
+use crate::serde::SERDE_AS_ION;
+use crate::{Decimal, Element};
+use serde::de::Error;
+use serde::{self, de, Deserialize, Serialize};
+use std::fmt;
+
+pub(crate) const ION_DECIMAL: &str = "$__ion_rs_decimal__";
+
+/// Serialization for Ion `Decimal`
+/// This serialization internally uses `serialize_newtype_struct` to trick serde to serialize a number value into decimal.
+/// This `newtype_struct` is named with `$__ion_rs_decimal__` to distinguish it from an actual `newtype_struct`.
+/// More information on `newtype_struct` can be found in the serde data model: `<https://serde.rs/data-model.html#types>`
+impl Serialize for Decimal {
+    #[inline]
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        SERDE_AS_ION.with(move |cell| {
+            let decimal: Decimal = self.clone();
+            if cell.get() {
+                serializer.serialize_newtype_struct(ION_DECIMAL, decimal.to_string().as_str())
+            } else {
+                serializer.serialize_str(decimal.to_string().as_str())
+            }
+        })
+    }
+}
+
+/// Deserialization for Ion `Decimal`
+/// This deserialization internally uses `serialize_newtype_struct` to trick serde to deserialize a number value into decimal.
+/// This `newtype_struct` is named with `$__ion_rs_decimal__` to distinguish it from an actual `newtype_struct`.
+/// More information on `newtype_struct` can be found in the serde data model: `<https://serde.rs/data-model.html#types>`
+impl<'de> Deserialize<'de> for Decimal {
+    fn deserialize<D>(deserializer: D) -> Result<Decimal, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        SERDE_AS_ION.with(move |cell| {
+            if cell.get() {
+                struct DecimalVisitor;
+
+                impl<'de> de::Visitor<'de> for DecimalVisitor {
+                    type Value = Decimal;
+
+                    fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                        formatter.write_str("an Ion Decimal")
+                    }
+
+                    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+                    where
+                        E: Error,
+                    {
```

Comment:
I expected `DecimalVisitor`'s implementation of `de::Visitor` to override the definition of [`visit_newtype_struct`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_newtype_struct) instead of `visit_str`. 

See [this snippet from a `serde` GH issue](https://github.com/serde-rs/serde/issues/2106#issue-1031792899).

Reply:
Removed `visit_str` from both decimal and timestamp deserializers as now we are using [transmute_copy](https://doc.rust-lang.org/std/mem/fn.transmute_copy.html) to treat a timestamp/decimal as `V::Value` which is visitor's value data type as defined in deserialization methods.